lib/ace/ext/papeeria/katex-previewer.coffee

```diff
@@ -242,148 +235,69 @@ define((require, exports, module) ->
+      curRow = row
+      curColumn = column
+
+      while true
+        nextIndex = curIndex + summand
+        # In case we're going backwards and have gone beyond the start of
+        # 'startRow'
+        if nextIndex < 0
+          startRow -= 1
+          curIndex = session.doc.positionToIndex({ row, column }, startRow)
+          nextIndex = curIndex + summand
+
+        { row: nextRow, column: nextColumn } = session.doc.indexToPosition(
+          nextIndex, startRow
+        )
+
+        # That means we're on the last row and last column
+        if nextColumn == curColumn and nextRow == curRow
+          return {
+            correct: false
+            reason: EquationRangeHandler.DOCUMENT_END_ERROR_CODE
+            row: curRow
+            column: curColumn
+          }
+
+        if getContext(session, nextRow, nextColumn) != EQUATION_CONTEXT
+          token = session.getTokenAt(curRow, curColumn)
+          if not token?
+            correct = false
+            reason = EquationRangeHandler.EMPTY_LINE_ERROR_CODE
+          else if isType(token, ERROR_TOKENTYPE)
+            correct = false
+            reason = EquationRangeHandler.WHITESPACE_LINE_ERROR_CODE
+          else
+            correct = true
+            reason = null
 
-      currentToken = tokenIterator.getCurrentToken()
-      prevRow = tokenIterator.getCurrentTokenPosition().row
-      # TODO: magic string? importing is hard though
-      boundaryCorrect = true
-      reasonCode = null
-      while LatexParsingContext.isType(currentToken, "equation")
-        currentToken = moveToBoundary()
-        if not currentToken?
-          boundaryCorrect = false
-          reasonCode = "js.math_preview.error.document_end"
-          break
-        currentRow = tokenIterator.getCurrentTokenPosition().row
-        # Empty string always means that equation state is popped from state stack.
-        # Unfortunately, empty string is not tokenized at all, and TokenIterator
-        # just skips it altogether, so we have to handle this manually here.
-        if Math.abs(currentRow - prevRow) > 1
-          boundaryCorrect = false
-          reasonCode = "js.math_preview.error.empty_line"
-          break
-        prevRow = currentRow
-
-      if currentToken? and LatexParsingContext.isType(currentToken, "error")
-        boundaryCorrect = false
-        reasonCode = "js.math_preview.error.whitespace_line"
-
-      moveFromBoundary()
-
-      { row: curTokenRow, column: curTokenColumn } = tokenIterator.getCurrentTokenPosition()
-      curTokenLength = tokenIterator.getCurrentToken().value.length
-      return {
-        correct: boundaryCorrect
-        reason: reasonCode
-        row: curTokenRow
-        column: curTokenColumn + (if start then 0 else curTokenLength)
-      }
+          return {
+            correct: correct
+            reason: reason
+            row: curRow
+            column: curColumn
+          }
+
+        curIndex = nextIndex
```

Comment:
Is it unreachable code? It seems that we never break the loop, we only return

Reply:
It is true we never break the loop, we only return. But this code is reachable, it is executed when we're still in an equation context.

Reply:
We can't get here without entring the loop. And we don't break the loop, we only return. Am I missing anything?

Reply:
We're still inside the loop here.

Reply:
Ah, I see, thanks